File: CSharp/ByRound/2042/2042_B_Game_with_Colored_Marbles.cs
// Problem: CF 2042 B - Game with Colored Marbles
// https://codeforces.com/contest/2042/problem/B

/*
B. Game with Colored Marbles
Algorithm: Greedy + Hash Map
Time Complexity: O(n) per test case
Space Complexity: O(n) for the hash map storing counts

The problem involves two players (Alice and Bob) taking turns removing marbles from a game.
Alice wants to maximize her score, Bob wants to minimize it. Both play optimally.

Alice's score is calculated as:
1 point for each color where she took at least one marble.
1 additional point for each color where she took ALL marbles of that color.

The key idea:
- For each color, we count how many marbles there are.
- We then determine whether Alice can secure all or just some of the marbles of a given color.
- If a color has only one marble, Alice might get 1 point (if she takes it) or 2 points if Bob takes the last one (and Alice gets both).
- The approach computes how many points each player will get based on optimal play.

The formula used:
ans = (c1 + 1) / 2 * 2 + dc - c1
Where:
- dc: number of distinct colors
- c1: count of colors with exactly one marble

This simplifies the logic for optimal greedy play.
*/

using System.Collections.Generic;
using System.Linq;
using System;
public class hello {
  static void Main() {
    var T = int.Parse(Console.ReadLine().Trim());
    while (T-- > 0) {
      Console.ReadLine();
      string[] line = Console.ReadLine().Trim().Split(' ');
      var a = Array.ConvertAll(line, int.Parse);
      getAns(a);
    }
  }
  static void getAns(int[] a) {
    var d = new Dictionary<int, int>(); // Store count of each color
    foreach (var x in a) {
      if (d.ContainsKey(x))
        d[x]++;
      else
        d[x] = 1;
    }
    var dc = d.Count; // Total number of distinct colors
    var c1 = d.Count(x => x.Value == 1); // Number of colors with exactly one marble
    var ans = (c1 + 1) / 2 * 2 + dc - c1; // Compute Alice's score using derived formula
    Console.WriteLine(ans);
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2042/2042_B_Game_with_Colored_Marbles.py
# Problem: CF 2042 B - Game with Colored Marbles
# https://codeforces.com/contest/2042/problem/B

# ******************************************************************************
# Problem: B. Game with Colored Marbles
# Algorithm: Game Theory + Greedy Analysis
# Time Complexity: O(n) per test case, where n is the number of marbles
# Space Complexity: O(n) for storing color counts and set operations
# 
# Approach:
# - For each test case:
#   1. Count occurrences of each color.
#   2. Determine how many colors have exactly one marble (Alice gets 1 point for taking at least one).
#   3. Determine how many colors have more than one marble (Alice might get points for taking all, but it's not always optimal).
#   4. The key insight is:
#      - If a color appears once: Alice gets 1 point if she takes the only marble.
#      - If a color appears more than once: She gets 1 point if she takes all marbles of that color.
#   5. Since Alice goes first and both play optimally, we simulate the optimal outcome:
#      - All colors with count 1 contribute to her points (she can take one).
#      - For each color appearing more than once, Alice can take all if she goes first in a round.
#      - The final score is based on how many distinct colors Alice ends up covering,
#        plus bonuses for full coverage of some colors.
#
# Note:
# This implementation simplifies the problem by focusing on counting.
# It assumes that optimal play results in maximizing her advantage from unique counts.
# ******************************************************************************

for s in [*open(0)][2::2]:  # Process odd lines (test case data), 0-indexed
    a = s.split()  # Get list of colors for current test case
    a = (*map(a.count, {*a}),)  # Count occurrences of each unique color and form tuple
    print(len(a) + a.count(1) % 2)  # Output computed score


# https://github.com/VaHiX/codeForces/